dainemo/autograd/ops/pool.mojo

```diff
@@ -0,0 +1,146 @@
+from dainemo.autograd.ops.conv import get_result_shape
+from dainemo.utils.tensorutils import calculate_strides
+
+
+# <------------MAXPOOL2D------------>
+struct MAXPOOL2D:
+    @staticmethod
+    fn forward[
+        kernel_shape: TensorShape,
+        padding: StaticIntTuple[2] = 0,
+        stride: StaticIntTuple[2] = 1,
+        dilation: StaticIntTuple[2] = 1
+    ](inputs: Node[dtype]) -> Node[dtype]:
+        """
+        Returns the max value of each kernel in the input tensor.
+            inputs.shape     [batch_size, in_channels, iX, iY]
+            kernel.shape     [out_channels, in_channels, kX, kY] 
+            outputs.shape    [batch_size, out_channels, oX, oY]
+            and for maxpool2d (in_channels == out_channels).
+        """
+
+        alias nelts: Int = simdwidthof[dtype]()
+        
+        let result_shape = get_result_shape[padding, stride, dilation](
+            inputs.tensor.shape(), kernel_shape
+        )
+
+        var outputs = Tensor[dtype](
+            inputs.tensor.dim(0), kernel_shape[0], result_shape[0], result_shape[1]
+        )
+        let outputs_strides = calculate_strides(outputs.shape())
+
+        for batch in range(inputs.tensor.dim(0)):
+            for in_ch in range(inputs.tensor.dim(1)):
+                for x in range(outputs.dim(2)):
+                    for y in range(outputs.dim(3)):
+                        var max_val: SIMD[dtype, 1] = -1e9
+                        let ix_base = x * stride[0] - padding[0]
+                        let iy_base = y * stride[1] - padding[1]
+                        for kx in range(kernel_shape[2]):
+                            for ky in range(kernel_shape[3]):
+                                let ix = ix_base + kx * dilation[0]
+                                let iy = iy_base + ky * dilation[1]
+
+                                if (
+                                    ix < 0
+                                    or iy < 0
+                                    or ix >= inputs.tensor.dim(2)
+                                    or iy >= inputs.tensor.dim(3)
+                                ):
+                                    continue
+
+                                let idx = (
+                                    batch * inputs.strides[0]
+                                    + in_ch * inputs.strides[1]
+                                    + ix * inputs.strides[2]
+                                    + iy
+                                )
+
+                                let val = inputs.tensor[idx]
+                                if val > max_val:
+                                    max_val = val
+
+                        let out_idx = (
+                            batch * outputs_strides[0]
+                            + in_ch * outputs_strides[1]
+                            + x * outputs_strides[2]
+                            + y
+                        )
+
+                        outputs[out_idx] = max_val
```

Comment:
Extra: And If you want I think this could be changed to be implemented in the same way as the conv2d function, to be able to remove the if inside the fors using the border calculations

Reply:
Yes, and probably we can also avoid those if's in the backward pass for both maxpooling and conv2d. Was thinking to implement this later indeed (after MNIST and when looking at vectorization/parallelization). But very nice you figured out the borders in the conv2d forward pass, so we have an example on how to do it.

Reply:
If good for you, can be merged